Userland/Libraries/LibGLSL/Parser.cpp

```diff
@@ -0,0 +1,1170 @@
+    decl->set_end(position());
+    return decl;
+}
+
+NonnullRefPtr<Expression const> Parser::parse_boolean_literal(ASTNode const& parent)
+{
+    auto token = consume(Token::Type::Keyword);
+    auto text = token.text();
+    // text == "true" || text == "false";
+    bool value = (text == "true");
+    return create_ast_node<BooleanLiteral>(parent, token.start(), token.end(), value);
+}
+
+bool Parser::match_boolean_literal()
+{
+    auto token = peek();
+    if (token.type() != Token::Type::Keyword)
+        return false;
+    auto text = token.text();
+    return text == "true" || text == "false";
+}
+
+NonnullRefPtr<Expression const> Parser::parse_numeric_literal(GLSL::ASTNode const& parent)
+{
+    auto token = consume();
+    auto text = token.text();
+    return create_ast_node<NumericLiteral>(parent, token.start(), token.end(), text);
+}
+
+bool Parser::match_numeric_literal()
+{
+    return match(Token::Type::Float) || match(Token::Type::Integer);
+}
+
+bool Parser::match_storage_qualifier()
+{
+    return match_keyword("const")
+        || match_keyword("in")
+        || match_keyword("out")
+        || match_keyword("inout")
+        || match_keyword("centroid")
+        || match_keyword("patch")
+        || match_keyword("sample")
+        || match_keyword("uniform")
+        || match_keyword("buffer")
+        || match_keyword("shared")
+        || match_keyword("coherent")
+        || match_keyword("volatile")
+        || match_keyword("restrict")
+        || match_keyword("readonly")
+        || match_keyword("writeonly")
+        || match_keyword("subroutine");
+}
+
+StorageTypeQualifier Parser::consume_storage_qualifier()
+{
+    VERIFY(peek().type() == Token::Type::Keyword);
+    auto keyword = consume().text();
+    if (keyword == "const")
+        return StorageTypeQualifier::Const;
+    if (keyword == "in")
+        return StorageTypeQualifier::In;
+    if (keyword == "out")
+        return StorageTypeQualifier::Out;
+    if (keyword == "inout")
+        return StorageTypeQualifier::Inout;
+    if (keyword == "centroid")
+        return StorageTypeQualifier::Centroid;
+    if (keyword == "patch")
+        return StorageTypeQualifier::Patch;
+    if (keyword == "sample")
+        return StorageTypeQualifier::Sample;
+    if (keyword == "uniform")
+        return StorageTypeQualifier::Uniform;
+    if (keyword == "buffer")
+        return StorageTypeQualifier::Buffer;
+    if (keyword == "shared")
+        return StorageTypeQualifier::Shared;
+    if (keyword == "coherent")
+        return StorageTypeQualifier::Coherent;
+    if (keyword == "volatile")
+        return StorageTypeQualifier::Volatile;
+    if (keyword == "restrict")
+        return StorageTypeQualifier::Restrict;
+    if (keyword == "readonly")
+        return StorageTypeQualifier::Readonly;
+    if (keyword == "writeonly")
+        return StorageTypeQualifier::Writeonly;
+    if (keyword == "subroutine")
+        return StorageTypeQualifier::Subroutine;
```

Comment:
For easy visual lookups, it might be beneficial to put these in alphabetical order.